Pass WSA_FLAG_NO_HANDLE_INHERIT flag to the WSASocketA() to avoid handle leaking on the Windows in case of using CreateProcess() on the server

frazenshtein opened the pull request:
https://docs.microsoft.com/en-us/windows/desktop/api/winsock2/nf-winsock2-wsasocketa

thelinuxfoundation commented:
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @frazenshtein isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

yang-g commented:
Please sign cla.

jtattermusch commented:
@frazenshtein   is this PR an attempt to resolve https://github.com/grpc/grpc/issues/17704?
Have you been able to confirm that it fixes the problem?

jtattermusch commented on src/core/lib/iomgr/tcp_server_windows.cc:
```diff
@@ -255,7 +255,7 @@ static grpc_error* start_accept_locked(grpc_tcp_listener* port) {
 
   sock = WSASocket(AF_INET6, SOCK_STREAM, IPPROTO_TCP, NULL, 0,
-                   WSA_FLAG_OVERLAPPED);
+                   WSA_FLAG_OVERLAPPED | WSA_FLAG_NO_HANDLE_INHERIT);
```
It looks like there are more places where WSASocket is created. Any reason why you're only updating occurrences in this file?

https://github.com/grpc/grpc/search?q=WSASocket&unscoped_q=WSASocket

frazenshtein replied:
I'm not sure of the semantics and mechanics used elsewhere and have only tested this server-side diff with some long-running stress tests. I think it's a good idea to set this flag in src/core/lib/iomgr/tcp_client_windows.cc and other places unless they imply using some kind of magic to pass a socket to child processes for some reason. But I'm afraid I won't be able to verify them properly. 
I can update the review placing WSA_FLAG_NO_HANDLE_INHERIT everywhere, if you rely on automated tests.

frazenshtein replied:
> @frazenshtein is this PR an attempt to resolve #17704?
> Have you been able to confirm that it fixes the problem?

I believe this PR will fix this issue. In my case, the problem looked like an error on the client side: `grpc on_accept error: the I/o operation was aborted due to the completion of the command flow or the request of the application.`

frazenshtein replied:
Btw this flag is used by default in the go: https://golang.org/src/net/sock_windows.go line 21

frazenshtein commented:
> @frazenshtein is this PR an attempt to resolve #17704?
> Have you been able to confirm that it fixes the problem?

I believe this PR will fix this issue. In my case, the problem looked like an error on the client side: `grpc on_accept error: the I/o operation was aborted due to the completion of the command flow or the request of the application.`

jtattermusch commented:
@nicolasnoble  opinions?

jtattermusch commented on src/core/lib/iomgr/socket_windows.cc:
```diff
@@ -181,4 +181,24 @@ int grpc_ipv6_loopback_available(void) {
+     https://msdn.microsoft.com/en-us/library/windows/desktop/ms742212(v=vs.85).aspx
+     for details. */
+  if (sock == INVALID_SOCKET && is_wsa_no_handle_inherit_set) {
+    g_is_wsa_no_handle_inherit_supported = false;
```
this is not threadsafe, we need a better way of detecting if WSA_FLAG_NO_HANDLE_INHERIT is supported.

frazenshtein replied:
Can you please suggest the best place to check that the host supports this flag? grpc_iomgr_init()?

apolcyn commented on test/cpp/naming/resolver_component_test.cc:
```diff
@@ -279,8 +279,8 @@ void OpenAndCloseSocketsStressLoop(int dummy_port, gpr_event* done_ev) {
-      SOCKET s = WSASocket(AF_INET6, SOCK_STREAM, IPPROTO_TCP, nullptr, 0,
-                           WSA_FLAG_OVERLAPPED);
+      SOCKET s = grpc_create_wsa_socket(AF_INET6, SOCK_STREAM, IPPROTO_TCP,
+                                        nullptr, 0, grpc_wsa_socket_flags);
```
nit: let's keep the test simple and as is.... this test doesn't have a use for `WSA_FLAG_NO_HANDLE_INHERIT`

jtattermusch commented:
@AspirinSJL  I think the general idea is of this PR is alright, can you please review for correctness from C-core perspective?

jtattermusch commented on src/core/lib/iomgr/socket_windows.cc:
```diff
@@ -39,6 +39,8 @@
 #include "src/core/lib/iomgr/sockaddr_windows.h"
 #include "src/core/lib/iomgr/socket_windows.h"
 
+DWORD grpc_wsa_socket_flags;
```
shouldn't there be a function to return the default flags to use?

apolcyn replied:
+1 - let's turn this into a function rather than have everyone reference the global directly

markdroth commented:
I think @apolcyn can review this instead of @AspirinSJL.  He has more familiarity with both Windows and with the c-ares code.

apolcyn commented:
Seeing this error in the linux artifact build:

```
[CXX]     Compiling src/core/lib/gpr/wrap_memcpy.cc
[CXX]     Compiling src/core/lib/gprpp/fork.cc
[CXX]     Compiling src/core/lib/gprpp/thd_posix.cc
[CXX]     Compiling src/core/lib/gprpp/thd_windows.cc
[CXX]     Compiling src/core/lib/profiling/basic_timers.cc
[AR]      Creating /tmp/libs/opt/libaddress_sorting.a
[CXX]     Compiling src/core/lib/profiling/stap_timers.cc
[AR]      Creating /tmp/libs/opt/libz.a
[AR]      Creating /tmp/libs/opt/libares.a
src/core/lib/iomgr/socket_windows.cc: In function 'void grpc_wsa_socket_flags_init()':
src/core/lib/iomgr/socket_windows.cc:193:77: error: passing NULL to non-pointer argument 5 of 'SOCKET WSASocketW(int, int, int, LPWSAPROTOCOL_INFOW, GROUP, DWORD)' [-Werror=conversion-null]
                           grpc_wsa_socket_flags | WSA_FLAG_NO_HANDLE_INHERIT);
```

apolcyn commented:
is the header that contains the new socket option also available to mingw? That build failure is a mingw issue

apolcyn commented:
I'm also wondering has this been verified to fix the issue. A test would be best but barring that it would be good to know that this has been manually verified

jtattermusch commented:
@frazenshtein, are you planning to address review feedback?

frazenshtein commented:
> @frazenshtein, are you planning to address review feedback?
Yes, I will try to take into account the comments in the nearest future. I have some difficulties with the allocation of free time.

frazenshtein commented:
> Seeing this error in the linux artifact build:

Yes, I really mixed up NULL and 0. Fixed.

> is the header that contains the new socket option also available to mingw? That build failure is a mingw issue 

Haven't found this one. Looks strange, because we have added #ifndef to the src/core/lib/iomgr/socket_windows.h Can you provide an error message or a link and keywords to grep?

> let's turn this into a function rather than have everyone reference the global directly

Fixed. However, I'm happy to rename it to your liking. I didn't find a good example in the code, so it turned out a bit long.

> I'm also wondering has this been verified to fix the issue. A test would be best but barring that it would be good to know that this has been manually verified

I was unable to reproduce the problem from the #17704 issue. If the author provides a minimal example it will be easier.

nicolasnoble commented on src/core/lib/iomgr/socket_windows.h:
```diff
@@ -114,6 +118,12 @@ void grpc_socket_become_ready(grpc_winsocket* winsocket,
    The value is probed once, and cached for the life of the process. */
 int grpc_ipv6_loopback_available(void);
 
+extern DWORD grpc_wsa_socket_flags;
```
Why having it both as a global extern and as a getter? We probably only want the getter. In which case, the variable should become static, and renamed to `s_wsa_socket_flags`

frazenshtein replied:
No hidden catch. No strings attached. Just my inattentiveness.

jtattermusch commented:
I triggered an adhoc package build (the nugets) to be able to verify if this patch fixes to do #17704 
https://sponge.corp.google.com/target?id=1c772a91-2b04-42ce-8d55-31a2a2990574&target=grpc/core/experimental/grpc_build_artifacts_multiplatform
I'll later post the link to the pre-built nugets.

jtattermusch commented:
The adhoc build of packages is here: 
https://storage.googleapis.com/grpc-testing-kokoro-prod/test_result_public/prod/grpc/core/master/windows/grpc_build_packages/5341/20190314-085942/github/grpc/artifacts/csharp_nugets_windows_dotnetcli.zip

jtattermusch commented:
According to https://github.com/grpc/grpc/issues/17704#issuecomment-481264974 this would fix #17704 if merged.

jtattermusch commented:
Do we want to use the default flags for https://github.com/grpc/grpc/blob/934e31d7d1c693348b10154b49fa863b48ed62ce/test/cpp/naming/resolver_component_test.cc#L278 too?

frazenshtein commented:
Could you please make clear for me is something expected from my side?

jtattermusch commented:
I believe we need need to use `grpc_get_default_wsa_socket_flags()` here:
https://github.com/grpc/grpc/blob/618a3f561d4a93f263cca23abad086ed8f4d5e86/src/core/lib/iomgr/tcp_server_windows.cc#L258

and here
https://github.com/grpc/grpc/blob/ad4f80569260f03c83fa7b372f50470c119dbb8a/test/cpp/naming/resolver_component_test.cc#L279

jtattermusch commented:
@apolcyn  I believe this is good to do, can you give owners approval so we can merge?
